File: 4119-minimum-distance-between-three-equal-elements-ii/4119-minimum-distance-between-three-equal-elements-ii.java
class Solution {
    public int minimumDistance(int[] nums) {
        int n = nums.length;
        int mini = Integer.MAX_VALUE;
        HashMap<Integer, ArrayList<Integer>> map = new HashMap<>();
        for (int i = 0; i < n; i++) {
            int current = nums[i];
            if (!map.containsKey(current))
                map.put(current, new ArrayList<>());
            map.get(current).add(i);
        }
        for (Map.Entry<Integer, ArrayList<Integer>> curr : map.entrySet()) {
            ArrayList<Integer> res = new ArrayList<>();
            res = curr.getValue();
            if (res.size() < 3) continue;
            for (int i = 0; i < res.size() - 2; i++) {
                mini = Math.min(mini, Math.abs(res.get(i) - res.get(i + 1)) + Math.abs(res.get(i + 1) - res.get(i + 2))+ Math.abs(res.get(i + 2) - res.get(i)));
            }
        }
        return mini == Integer.MAX_VALUE ? -1 : mini;
    }
}

File: 4119-minimum-distance-between-three-equal-elements-ii/4119-minimum-distance-between-three-equal-elements-ii.py
class Solution:
    def minimumDistance(self, nums: List[int]) -> int:
        n = len(nums)
        map = dict()

        for i in range(n):
            if nums[i] in map:
                map[nums[i]].append(i)
            else:
                map[nums[i]] = []
                map[nums[i]].append(i)

        mini = float("inf")
        for key, val in map.items():
            if len(val) < 3:
                continue
            for i in range(len(val) - 2):
                mini = min(
                    mini,
                    abs(val[i] - val[i + 1])
                    + abs(val[i + 1] - val[i + 2])
                    + abs(val[i + 2] - val[i]),
                )

        if mini == float("inf"):
            return -1
        return (int)(mini)

